lib/api.js

```diff
@@ -0,0 +1,213 @@
+var fork = require('./fork');
+
+function Api(options) {
+	if (!(this instanceof Api)) {
+		return new Api(options);
+	}
+
+	EventEmitter.call(this);
+
+	assign(this, options);
+
+	this.rejectionCount = 0;
+	this.exceptionCount = 0;
+	this.passCount = 0;
+	this.failCount = 0;
+	this.fileCount = 0;
+	this.testCount = 0;
+	this.errors = [];
+	this.stats = [];
+	this.tests = [];
+
+	Object.keys(Api.prototype).forEach(function (key) {
+		this[key] = this[key].bind(this);
+	}, this);
+}
+
+util.inherits(Api, EventEmitter);
+module.exports = Api;
+
+Api.prototype._runFile = function (file) {
+	var args = [file];
+
+	if (this.failFast) {
+		args.push('--fail-fast');
+	}
+
+	if (this.serial) {
+		args.push('--serial');
+	}
+
+	// Forward the `time-require` `--sorted` flag.
+	// Intended for internal optimization tests only.
+	if (this._sorted) {
+		args.push('--sorted');
+	}
+
+	return fork(args)
+		.on('stats', this._handleStats)
+		.on('test', this._handleTest)
+		.on('unhandledRejections', this._handleRejections)
+		.on('uncaughtException', this._handleExceptions);
+};
+
+Api.prototype._handleRejections = function (data) {
+	this.emit('rejection', data);
+	this.rejectionCount += data.rejections.length;
+};
+
+Api.prototype._handleExceptions = function (data) {
+	this.emit('exception', data);
+	this.exceptionCount++;
+};
+
+Api.prototype._handleStats = function (stats) {
+	this.testCount += stats.testCount;
+};
+
+Api.prototype._handleTest = function (test) {
+	test.title = this._prefixTitle(test.file) + test.title;
+
+	var isError = test.error.message;
+
+	if (isError) {
+		this.errors.push(test);
+	} else {
+		test.error = null;
+	}
+
+	this.emit('test', test);
+};
+
+Api.prototype._prefixTitle = function (file) {
+	if (this.fileCount === 1) {
+		return '';
+	}
+
+	var separator = ' ' + chalk.gray.dim(figures.pointerSmall) + ' ';
+
+	var base = path.dirname(this.files[0]);
+
+	if (base === '.') {
+		base = this.files[0] || 'test';
+	}
+
+	base += path.sep;
+
+	var prefix = path.relative('.', file)
+		.replace(base, '')
+		.replace(/\.spec/, '')
+		.replace(/test\-/g, '')
+		.replace(/\.js$/, '')
+		.split(path.sep)
+		.join(separator);
+
+	if (prefix.length > 0) {
+		prefix += separator;
+	}
+
+	return prefix;
+};
+
+Api.prototype.run = function () {
+	var self = this;
+
+	return handlePaths(this.files)
+		.map(function (file) {
+			return path.resolve(file);
+		})
+		.then(function (files) {
+			if (files.length === 0) {
+				var err = new Error('Couldn\'t find any files to test');
+				return Promise.reject(err);
+			}
+
+			self.fileCount = files.length;
+
+			var tests = files.map(self._runFile);
+
+			// receive test count from all files and then run the tests
+			var statsCount = 0;
+			var deferred = Promise.pending();
+
+			tests.forEach(function (test) {
+				test.on('stats', function () {
+					if (++statsCount === self.fileCount) {
+						self.emit('ready');
+
+						tests.forEach(function (test) {
+							test.run();
```

Comment:
Why are you waiting for every test file to load before kicking off the tests? Once an individual test is `ready`, why not let it start?